### ISSUE TITLE
Use Software Heritage as a fallback download location

### ISSUE THREAD

**Issue by armijnhemel** [NixOS/nixpkgs#53653]
## Issue description

In the past download locations for software have disappeared for various reasons: source releases were deleted, domain name registration expired, website redesign, and so on. This has led to packages failing to build.

Although there are sometimes mirrors of source code these might come with the same issue, for example when they are exact clones of the original site.

It might be good to use Software Heritage as a fallback download location, or at least as a source in case we want to provide our own mirror of the source code that disappeared from the original site. Currently Software Heritage already indexes everything from Debian, GitHub, GitLab, PyPI, Google Code, GNU and several others and make it available under a unique hash

### Steps to reproduce

None

## Technical details

None

**Comment by armijnhemel** [comment]
Adding @edolstra

**Comment by armijnhemel** [comment]
Yes, you can get content from SH using various checksums. There is rate limiting in place though (but for for example Hydra that could possibly be lifted if needed).

**Comment by davidak** [comment]
I like the idea! Do they support such usage of their system?

**Comment by armijnhemel** [comment]
My guess is that it depends on how we implement it and we would need to talk to them about it. My guess is that if we do it just as a backup (as soon as the original is no longer available) and first limit it to Hydra that it would be no problem for Software Heritage to lift the rate limit.

**Comment by armijnhemel** [comment]
Guix implemented it recently: https://lwn.net/Articles/784401/

**Comment by asymmetric** [comment]
And a direct link to the [blog post](https://www.gnu.org/software/guix/blog/2019/connecting-reproducible-deployment-to-a-long-term-source-code-archive/).

**Comment by makefu** [comment]
Imho "Debian, GitHub, GitLab, PyPI, Google Code, GNU" are not the pain points we have with nixpkgs right now. The biggest issue i see are mirrors of nonfree software such as Oracle, Nvidia, AMD or dropbox. Sources like that are no issue for GUIX due to the free nature of the repository packages.

I really like the idea of mirroring all the sources we use but maybe we can get help from other projects or technologies (ipfs, archive.org) which do not have the restriction to only mirror free source code.

**Comment by davidak** [comment]
@makefu right. for me, steam, amd and adobe flash are the only issues of this kind. I think most of the time it would be a licensing issue to redistribute nonfree software.

But it would still be very elegant to use Software Heritage in the longterm perspective. It might saves you some day. Imagine you want to open an obscure file format popular today in 50 years. When we have implemented it NOW, you will be able to use todays nixpkgs then.

That's to different problems to solve.

**Comment by asymmetric** [comment]
> Imho "Debian, GitHub, GitLab, PyPI, Google Code, GNU" are not the pain points we have with nixpkgs right now.

@makefu They are not the problem right now, but they're guaranteed to be one in the future, especially the commercial companies in that list.

Given that all of those repositories will eventually go down, isn't it better to have as fallback an institution that explicitly has the purpose of long-term conservancy?

I agree that a decentralized storage layer like IPFS would also be very interesting. But maybe that's the topic for another issue?

**Comment by makefu** [comment]
I am totally for implementing software heritage as backend, i just wanted to point out that the things currently mirrored are only a subset of the sources we have (and break) in nixpkgs.

The nice thing with the current setup in nixpkgs is that we can just add more options.

**Comment by davidak** [comment]
>I agree that a decentralized storage layer like IPFS would also be very interesting. But maybe that's the topic for another issue?

It is. https://github.com/NixOS/nix/issues/859

**Comment by seirl** [comment]
@makefu :

> I really like the idea of mirroring all the sources we use but maybe we can get help from other projects or technologies (ipfs, archive.org) which do not have the restriction to only mirror free source code.

We do not have that restriction in Software Heritage. You're free to deposit code with a non-free license. It's just easier for us to mirror the main forges in priority, because that's where most of the code is available.

**Comment by makefu** [comment]
@seirl thanks for the reply! That is great to hear.
Is it also possible to mirror blobs (e.g. VirtualBox Extensions or Adobe Flash Player binary)?

**Comment by seirl** [comment]
@makefu It's technically possible, but it's not the intent of Software Heritage to mirror binaries. We won't filter them, but be aware that there are size restrictions that will apply. If it's for the exception rather than the norm (e.g, one package that vendors a small proprietary .so) it seems totally fine, but please don't use Software Heritage as a binary cache :-)

Cheers.

**Comment by seirl** [comment]
Just to add, the size restriction is currently at 100 MiB, but it's not a hard guarantee and could change in the future. We expect most of the source files deposited to be *way* under that.

**Comment by makefu** [comment]
@seirl thanks for clearing that up, i am sure more people are interested in this response as well :+1:

**Comment by Zimmi48** [comment]
> Imho "Debian, GitHub, GitLab, PyPI, Google Code, GNU" are not the pain points we have with nixpkgs right now.

Just because GitHub is not expected to go down anytime soon doesn't prevent people from deleting their repository. Oh, and it also regularly happens to GitHub to be down / have troubles for a few hours.

**Comment by davidak** [comment]
This is still not merged.

**Comment by asymmetric** [comment]
AFAIK there's some ongoing work on this [here](https://github.com/nix-community/nixpkgs-swh), although I'm not sure of the current status. /cc @nlewo.

See also [this discourse post](https://discourse.nixos.org/t/tweag-nix-dev-update-2/6808?u=asymmetric).

**Comment by rgrunbla** [comment]
Regarding this, you can find the archives generated by hydra every day ( see https://nix-community.github.io/nixpkgs-swh/ ) encoded as "branches" on this repository of the software heritage website: https://archive.softwareheritage.org/browse/origin/directory/?origin_url=https://nix-community.github.io/nixpkgs-swh/sources-unstable.jso

**Comment by nixinator** [comment]
did the fallback ever get merged? maybe i missed a commit.

https://www.tweag.io/blog/2020-06-18-software-heritage/

**Comment by nlewo** [comment]
> did the fallback ever get merged? maybe i missed a commit.

AFIAK, the fallback has never been implemented (only some partial prototypes).

**Comment by progval** [comment]
By the way, Nixpkgs currently can't be archived by SWH because of https://github.com/nix-community/nixpkgs-swh/issues/5

**Comment by christoph-blessing** [comment]
I stumbled across this when looking into combining the source code from SWH with the build instructions from nixpkgs for a work project but it seems to be abandoned. Does anyone know what the current status is?
